traffic_portal/app/src/common/modules/table/profilesParamsCompare/TableProfilesParamsCompareController.js

```diff
@@ -0,0 +1,213 @@
+
+	this.profile1Usage;
+	this.profile2Usage;
+
+	let updateProfile1 = false,
+		updateProfile2 = false;
+
+	let getProfileUsage = function(profile, profNum) {
+		if (profile.type === 'DS_PROFILE') { // if this is a ds profile, then it is used by delivery service(s) so we'll fetch the ds count...
+			deliveryServiceService.getDeliveryServices({ profile: profile.id }).
+				then(function(result) {
+					this['profile' + profNum + 'Usage'] = result.length + ' delivery services';
+				});
+		} else { // otherwise the profile is used by servers so we'll fetch the server count...
+			serverService.getServers({ profileId: profile.id }).
+				then(function(result) {
+					this['profile' + profNum + 'Usage'] = result.length + ' servers';
+				});
+		}
+	};
+
+	let confirmUpdateProfiles = function() {
+		// ok, this method is fun :)
+		let params = {
+			title: 'Modify ' + profile1.name + ' parameters',
+			message: 'The ' + profile1.name + ' profile is used by ' + this.profile1Usage + '. Are you sure you want to modify the parameters?'
+		};
+		let modalInstance = $uibModal.open({
+			templateUrl: 'common/modules/dialog/confirm/dialog.confirm.tpl.html',
+			controller: 'DialogConfirmController',
+			size: 'md',
+			resolve: {
+				params: function () {
+					return params;
+				}
+			}
+		});
+		modalInstance.result
+			.then(
+				function() { updateProfile1 = true; },
+				function() { updateProfile1 = false; }
+			)
+			.finally(
+				function() {
+					let params = {
+						title: 'Modify ' + profile2.name + ' parameters',
+						message: 'The ' + profile2.name + ' profile is used by ' + this.profile2Usage + '. Are you sure you want to modify the parameters?'
+					};
+					let modalInstance = $uibModal.open({
+						templateUrl: 'common/modules/dialog/confirm/dialog.confirm.tpl.html',
+						controller: 'DialogConfirmController',
+						size: 'md',
+						resolve: {
+							params: function () {
+								return params;
+							}
+						}
+					});
+					modalInstance.result
+						.then(
+							function() { updateProfile2 = true; },
+							function() { updateProfile2 = false; }
+						)
+						.finally(
+							function() {
+								let promises = [];
+
+								if (updateProfile1) {
+									promises.push(profileParameterService.linkProfileParameters(profile1, _.pluck($scope.selectedProfile1Params, 'id')));
+								}
+
+								if (updateProfile2) {
+									promises.push(profileParameterService.linkProfileParameters(profile2, _.pluck($scope.selectedProfile2Params, 'id')));
+								}
+
+								if (promises.length > 0) {
+									$q.all(promises)
+										.then(
+											function(result) {
+												let messages = [];
+												for (let i = 0; i < result.length; i++) {
+													messages = _.union(messages, result[i].data.alerts);
```

Comment:
Instead of using `_.union`, consider using the built-in [`Set` class](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Set) like so:

```javascript
const messages = new Set(); // 'messages' isn't being reassigned, so it's constant
for (let r of result) {
	for (let a of r.data.alerts) {
		messages.add(a);
	}
}
// You can submit it as an array using Array.from:
messageModel.setMessages(Array.from(messages), false);
```
For bonus points, you can use [`Array.prototype.flatMap`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/flatMap) to flatten it into a one-dimensional array prior to iteration:
```javascript
const messages = new Set();
// This loop directly iterates every alert of every response
for (let a of result.flatMap(function(x){return x.data.alerts})) {
	messages.add(a);
}
messageModel.setMessages(messages, false);
```

Reply:
Error: Parsing file /traffic_portal/app/src/common/modules/table/profilesParamsCompare/TableProfilesParamsCompareController.js

i guess it didn't like the `of` loop or something so just stuck with boring old for loop